File: javascript/refactoring_destination.js
/**
 * JavaScript test file for cross-file refactoring destination.
 * 
 * This file contains modified versions of functions moved from refactoring_source.js
 * to test DiffScope's ability to detect cross-file refactorings.
 */

/**
 * Process a list of data objects and compute statistics with enhanced features.
 * @param {Array<Object>} data - List of objects containing data to process
 * @param {Object} [options] - Processing options
 * @param {boolean} [options.includeAdvancedStats=false] - Whether to include advanced statistical calculations
 * @param {Array<string>} [options.excludeKeys=[]] - Keys to exclude from analysis
 * @returns {Object} Object with computed statistics
 */
function processDataEnhanced(data, options = {}) {
  const { includeAdvancedStats = false, excludeKeys = [] } = options;
  
  if (!data || data.length === 0) {
    return { count: 0, hasData: false, error: null };
  }
  
  try {
    // Extract values by type from each item
    const numericValues = [];
    const textValues = [];
    const booleanValues = [];
    
    for (const item of data) {
      for (const [key, value] of Object.entries(item)) {
        if (excludeKeys.includes(key)) continue;
        
        if (typeof value === 'number' && !Number.isNaN(value)) {
          numericValues.push(value);
        } else if (typeof value === 'string') {
          textValues.push(value);
        } else if (typeof value === 'boolean') {
          booleanValues.push(value);
        }
      }
    }
    
    // Compute statistics
    const result = {
      count: data.length,
      hasData: data.length > 0,
      keysFound: [...new Set(data.flatMap(item => Object.keys(item)))],
      valueTypes: {
        numeric: numericValues.length,
        text: textValues.length,
        boolean: booleanValues.length
      }
    };
    
    if (numericValues.length > 0) {
      numericValues.sort((a, b) => a - b);
      const midpoint = Math.floor(numericValues.length / 2);
      const median = numericValues.length % 2 !== 0
        ? numericValues[midpoint]
        : (numericValues[midpoint - 1] + numericValues[midpoint]) / 2;
      
      const sum = numericValues.reduce((acc, val) => acc + val, 0);
      const mean = sum / numericValues.length;
      
      result.numericStats = {
        min: Math.min(...numericValues),
        max: Math.max(...numericValues),
        mean,
        median,
        sum,
        count: numericValues.length
      };
      
      if (includeAdvancedStats && numericValues.length > 1) {
        // Calculate variance and standard deviation
        const variance = numericValues.reduce((acc, val) => acc + Math.pow(val - mean, 2), 0) / numericValues.length;
        const stddev = Math.sqrt(variance);
        
        // Calculate quartiles
        const q1Index = Math.floor(numericValues.length * 0.25);
        const q3Index = Math.floor(numericValues.length * 0.75);
        
        result.numericStats.advancedStats = {
          variance,
          stddev,
          quartiles: {
            q1: numericValues[q1Index],
            q2: median,
            q3: numericValues[q3Index]
          }
        };
      }
    }
    
    if (booleanValues.length > 0) {
      const trueCount = booleanValues.filter(Boolean).length;
      result.booleanStats = {
        trueCount,
        falseCount: booleanValues.length - trueCount,
        truePercentage: (trueCount / booleanValues.length) * 100
      };
    }
    
    return result;
  } catch (error) {
    return {
      count: data.length,
      hasData: data.length > 0,
      error: error.message
    };
  }
}

/**
 * Transform a data structure based on a key mapping with enhanced options.
 * @param {Object|Array<Object>} data - Data structure to transform
 * @param {Object} mapping - Object mapping old keys to new keys
 * @param {Object} [options] - Transformation options
 * @param {boolean} [options.recursive=false] - Whether to apply transformation recursively
 * @param {Array<string>} [options.filterKeys=null] - Optional list of keys to include in output
 * @returns {Object|Array<Object>} Transformed data structure
 */
function transformDataStructure(data, mapping, options = {}) {
  const { recursive = false, filterKeys = null } = options;
  
  // Handle array case
  if (Array.isArray(data)) {
    return data
      .filter(item => typeof item === 'object' && item !== null)
      .map(item => transformDataStructure(item, mapping, options));
  }
  
  // Now data is guaranteed to be an object
  const result = {};
  
  // Apply mappings
  for (const [oldKey, newKey] of Object.entries(mapping)) {
    if (Object.prototype.hasOwnProperty.call(data, oldKey)) {
      let value = data[oldKey];
      
      // Recursive transformation
      if (recursive) {
        if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
          value = transformDataStructure(value, mapping, options);
        } else if (Array.isArray(value)) {
          value = transformDataStructure(value, mapping, options);
        }
      }
      
      result[newKey] = value;
    }
  }
  
  // Include unmapped keys
  for (const [key, value] of Object.entries(data)) {
    if (!Object.prototype.hasOwnProperty.call(mapping, key)) {
      let transformedValue = value;
      
      // Recursive transformation
      if (recursive) {
        if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
          transformedValue = transformDataStructure(value, mapping, options);
        } else if (Array.isArray(value)) {
          transformedValue = transformDataStructure(value, mapping, options);
        }
      }
      
      result[key] = transformedValue;
    }
  }
  
  // Filter keys if required
  if (Array.isArray(filterKeys)) {
    const filteredResult = {};
    for (const key of filterKeys) {
      if (Object.prototype.hasOwnProperty.call(result, key)) {
        filteredResult[key] = result[key];
      }
    }
    return filteredResult;
  }
  
  return result;
}

/**
 * Parse structured data from string with support for multiple formats.
 * @param {string} text - String to parse
 * @param {string} [formatType="key_value"] - Format type ("key_value", "json", etc.)
 * @returns {Object} Object with parsed data
 */
function parseStructuredData(text, formatType = "key_value") {
  if (!text) {
    return {};
  }
  
  if (formatType === "key_value") {
    // Split by semicolons, but respect quoted values
    const regex = /([^;=]+)=(?:"([^"]*)"|([^;]*))/g;
    const result = {};
    let match;
    
    while ((match = regex.exec(text)) !== null) {
      const key = match[1].trim();
      const value = match[2] !== undefined ? match[2] : match[3].trim();
      
      // Try to convert to appropriate type
      if (value.toLowerCase() === 'true') {
        result[key] = true;
      } else if (value.toLowerCase() === 'false') {
        result[key] = false;
      } else if (!isNaN(value) && value !== '') {
        if (value.includes('.')) {
          result[key] = parseFloat(value);
        } else {
          result[key] = parseInt(value, 10);
        }
      } else {
        result[key] = value;
      }
    }
    
    return result;
  } else if (formatType === "json") {
    try {
      return JSON.parse(text);
    } catch (error) {
      return { error: "Invalid JSON" };
    }
  } else {
    return { error: `Unsupported format: ${formatType}` };
  }
}

/**
 * Extract specific metrics from a complex data structure.
 * @param {Object} data - Complex data structure
 * @param {Array<string>} metricPaths - List of dot-notation paths to extract
 * @returns {Object} Object with extracted metrics
 */
function extractMetricData(data, metricPaths) {
  const result = {};
  
  for (const path of metricPaths) {
    const parts = path.split('.');
    let current = data;
    let isValid = true;
    
    try {
      for (const part of parts) {
        if (current === null || current === undefined || typeof current !== 'object') {
          isValid = false;
          break;
        }
        current = current[part];
      }
      
      result[path] = isValid ? current : null;
    } catch (error) {
      result[path] = null;
    }
  }
  
  return result;
}

// Test cases
if (require.main === module) {
  // Test processDataEnhanced
  const testData = [
    { name: "Item 1", value: 10, active: true },
    { name: "Item 2", value: 20, active: false },
    { name: "Item 3", value: 30, active: true }
  ];
  const stats = processDataEnhanced(testData, { includeAdvancedStats: true });
  console.log("Enhanced data statistics:", stats);
  
  // Test parseStructuredData
  const testString = 'name="Complex Item"; value=42; active=true; tags="javascript,test,example"';
  const parsed = parseStructuredData(testString, "key_value");
  console.log("Parsed structured data:", parsed);
  
  // Test transformDataStructure with options
  const testMapping = { name: "title", value: "amount", active: "isActive" };
  const transformed = transformDataStructure(testData, testMapping, { recursive: true });
  console.log("Transformed data:", transformed);
  
  // Test extractMetricData
  const complexData = {
    user: {
      profile: {
        name: "Test User",
        age: 30
      }
    },
    metrics: {
      visits: 10,
      conversions: 2
    }
  };
  const metrics = extractMetricData(complexData, ["user.profile.name", "metrics.visits"]);
  console.log("Extracted metrics:", metrics);
}

module.exports = {
  processDataEnhanced,
  parseStructuredData,
  transformDataStructure,
  extractMetricData
}; 

File: python/refactoring_destination.py
"""
Python test file for cross-file refactoring destination.

This file contains modified versions of functions moved from refactoring_source.py
to test DiffScope's ability to detect cross-file refactorings.
"""

import math
import re
import statistics
from typing import List, Dict, Any, Optional, Union, Tuple


def process_data_enhanced(data: List[Dict[str, Any]], include_advanced_stats: bool = False) -> Dict[str, Any]:
    """
    Process a list of data dictionaries and compute statistics with enhanced features.
    
    Args:
        data: List of dictionaries containing data to process
        include_advanced_stats: Whether to include advanced statistical calculations
        
    Returns:
        Dictionary with computed statistics
    """
    if not data:
        return {"count": 0, "has_data": False, "error": None}
    
    try:
        # Extract numeric values from each item
        numeric_values = []
        text_values = []
        boolean_values = []
        
        for item in data:
            for key, value in item.items():
                if isinstance(value, (int, float)) and not isinstance(value, bool):
                    numeric_values.append(value)
                elif isinstance(value, str):
                    text_values.append(value)
                elif isinstance(value, bool):
                    boolean_values.append(value)
        
        # Compute statistics
        result = {
            "count": len(data),
            "has_data": len(data) > 0,
            "keys_found": set().union(*(d.keys() for d in data)),
            "value_types": {
                "numeric": len(numeric_values),
                "text": len(text_values),
                "boolean": len(boolean_values)
            }
        }
        
        if numeric_values:
            result["numeric_stats"] = {
                "min": min(numeric_values),
                "max": max(numeric_values),
                "mean": sum(numeric_values) / len(numeric_values),
                "median": statistics.median(numeric_values),
                "sum": sum(numeric_values),
                "count": len(numeric_values)
            }
            
            if include_advanced_stats and len(numeric_values) > 1:
                result["numeric_stats"].update({
                    "stddev": statistics.stdev(numeric_values),
                    "variance": statistics.variance(numeric_values),
                    "quartiles": {
                        "q1": statistics.quantiles(numeric_values, n=4)[0],
                        "q2": statistics.quantiles(numeric_values, n=4)[1],
                        "q3": statistics.quantiles(numeric_values, n=4)[2]
                    }
                })
        
        if boolean_values:
            result["boolean_stats"] = {
                "true_count": sum(1 for v in boolean_values if v),
                "false_count": sum(1 for v in boolean_values if not v),
                "true_percentage": sum(1 for v in boolean_values if v) / len(boolean_values) * 100
            }
        
        return result
    
    except Exception as e:
        return {
            "count": len(data),
            "has_data": len(data) > 0,
            "error": str(e)
        }


def transform_data_structure(
    data: Union[Dict[str, Any], List[Dict[str, Any]]],
    mapping: Dict[str, str],
    recursive: bool = False,
    filter_keys: Optional[List[str]] = None
) -> Union[Dict[str, Any], List[Dict[str, Any]]]:
    """
    Transform a data structure based on a key mapping with enhanced options.
    
    Args:
        data: Data structure to transform (dictionary or list of dictionaries)
        mapping: Dictionary mapping old keys to new keys
        recursive: Whether to apply transformation recursively to nested structures
        filter_keys: Optional list of keys to include in the output (applied after mapping)
        
    Returns:
        Transformed data structure
    """
    # Handle the case of list of dictionaries
    if isinstance(data, list):
        return [transform_data_structure(item, mapping, recursive, filter_keys) 
                for item in data if isinstance(item, dict)]
    
    # Now data is guaranteed to be a dictionary
    result = {}
    
    for old_key, new_key in mapping.items():
        if old_key in data:
            value = data[old_key]
            
            # Apply recursive transformation if needed
            if recursive and isinstance(value, dict):
                value = transform_data_structure(value, mapping, recursive, filter_keys)
            elif recursive and isinstance(value, list):
                value = transform_data_structure(value, mapping, recursive, filter_keys)
                
            result[new_key] = value
    
    # Include any keys not in the mapping
    for key, value in data.items():
        if key not in mapping:
            # Apply recursive transformation if needed
            if recursive and isinstance(value, dict):
                value = transform_data_structure(value, mapping, recursive, filter_keys)
            elif recursive and isinstance(value, list):
                value = transform_data_structure(value, mapping, recursive, filter_keys)
                
            result[key] = value
    
    # Filter keys if specified
    if filter_keys:
        result = {k: v for k, v in result.items() if k in filter_keys}
    
    return result


def parse_structured_data(text: str, format_type: str = "key_value") -> Dict[str, Any]:
    """
    Parse structured data from string with support for multiple formats.
    
    Args:
        text: String to parse
        format_type: Format type to parse ("key_value", "json", "csv", etc.)
        
    Returns:
        Dictionary with parsed data
    """
    if not text:
        return {}
    
    if format_type == "key_value":
        # Split by semicolons, but respect quoted values
        pairs = re.findall(r'([^;=]+)=(?:"([^"]*)"|([^;]*))', text)
        
        result = {}
        for key, quoted_val, unquoted_val in pairs:
            key = key.strip()
            value = quoted_val if quoted_val else unquoted_val.strip()
            
            # Try to convert to appropriate type
            try:
                if value.lower() == 'true':
                    value = True
                elif value.lower() == 'false':
                    value = False
                elif '.' in value and value.replace('.', '').isdigit():
                    value = float(value)
                elif value.isdigit():
                    value = int(value)
            except (ValueError, AttributeError):
                pass
            
            result[key] = value
        
        return result
    elif format_type == "json":
        import json
        try:
            return json.loads(text)
        except json.JSONDecodeError:
            return {"error": "Invalid JSON"}
    else:
        return {"error": f"Unsupported format: {format_type}"}


def extract_metric_data(data: Dict[str, Any], metric_paths: List[str]) -> Dict[str, Any]:
    """
    Extract specific metrics from a complex data structure.
    
    Args:
        data: Complex data structure
        metric_paths: List of dot-notation paths to extract (e.g., "user.profile.age")
        
    Returns:
        Dictionary with extracted metrics
    """
    result = {}
    
    for path in metric_paths:
        parts = path.split('.')
        current = data
        
        try:
            for part in parts:
                current = current[part]
            
            result[path] = current
        except (KeyError, TypeError):
            result[path] = None
    
    return result


if __name__ == "__main__":
    # Test process_data_enhanced
    test_data = [
        {"name": "Item 1", "value": 10, "active": True},
        {"name": "Item 2", "value": 20, "active": False},
        {"name": "Item 3", "value": 30, "active": True}
    ]
    stats = process_data_enhanced(test_data, include_advanced_stats=True)
    print("Enhanced data statistics:", stats)
    
    # Test parse_structured_data
    test_string = 'name="Complex Item"; value=42; active=true; tags="python,test,example"'
    parsed = parse_structured_data(test_string, format_type="key_value")
    print("Parsed structured data:", parsed)
    
    # Test transform_data_structure with enhanced options
    test_mapping = {"name": "title", "value": "amount", "active": "is_active"}
    transformed = transform_data_structure(test_data, test_mapping, recursive=True)
    print("Transformed data:", transformed)
    
    # Test extract_metric_data
    complex_data = {
        "user": {
            "profile": {
                "name": "Test User",
                "age": 30
            }
        },
        "metrics": {
            "visits": 10,
            "conversions": 2
        }
    }
    metrics = extract_metric_data(complex_data, ["user.profile.name", "metrics.visits"])
    print("Extracted metrics:", metrics) 